### ISSUE TITLE
The 'gpt' counter reports that its index begins with the number 1 instead of 0

### ISSUE THREAD

**Issue by NickMRamirez** [haproxy/haproxy#1630]
### Detailed Description of the Problem

I have defined the following stick table in a `peers` section:

```
table response_times type binary len 20 size 1000000 store gpt(0)
```

According to the [2.5 documentation for a stick table](https://cbonte.github.io/haproxy-dconv/2.5/configuration.html#stick-table), the `gpt` counter should allow me to set its index to 0, as I have done above:

> - gpt(<nb>) : General Purpose Tags Array of <nb> elements. This is an array
    of positive 32-bit integers which may be used for anything.
    Most of the time they will be used to put a special tags on some entries,
    for instance to note that a specific behavior was detected and must be
    known for future matches. **This array is limited to a maximum of 100
    elements: gpt0 to gpt99**, to ensure that the build of a peer update
    message can fit into the buffer.

Here, `gpt0` to me indicates that I should be able to set the counter as `gpt(0)`. However, if I try this, I get an error upon startup:

> table: array size is out of allowed range (1-100) for store option 'gpt'.

This error indicates that I must use an index of 1. If I change the counter from `gpt(0)` to `gpt(1)`, the error goes away. 

Is the documentation incorrect?

### Expected Behavior

It is unclear what the expected behavior is, but from the docs, I assumed that I would be able to set `gpt(0)` (a zero-based index).

### Steps to Reproduce the Behavior

1. Create a table definition that stores `gpt(0)`
2. Reload HAProxy and you will see the error.

### Do you have an idea how to solve the issue?

The documentation may simply be wrong, or I may have misunderstood it.

### What is your configuration?

```haproxy
peers mypeers
  bind :10000
  server enterprise1
  server profiling-engine 127.0.0.1:10001

  # A 'gpt' is a 'general purpose tag', which you can set to an integer,
  # unlike a 'gpc' which you can increment only. The 'tag' parts of 
  # the name denotes its original use case, which was to tag certain
  # sessions (i.e. 0 or 1) to flag them for some reason. But, because
  # it is a 32-bit int, you can set it to anything. In this case,
  # we store the number of seconds it took for the client to get
  # a response.
  # The 'type' is 'binary'. We will store 'base32+src', which is a
  # hash of the Host header and the URL path, plus the client's source
  # IP address. This means that the key in the table is tracking
  # each client's response rate at a given URL path, which will give
  # us a broad picture of the responsiveness over the whole site.
  table response_times type binary len 20 size 1000000 store gpt(0)
```
```


### Output of `haproxy -vv`

```plain
HAProxy version 2.5.0-1.0.0-269.191 2022/03/10 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around November 2023.
Known bugs: https://www.haproxy.com/documentation/hapee/2-5r1/onepage/changelog/#2.5.0
Running on: Linux 5.4.0-88-generic #99-Ubuntu SMP Thu Sep 23 17:29:00 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -Werror -DMAX_SESS_STKCTR=12 -DSTKTABLE_EXTRA_DATA_TYPES=10
  OPTIONS = USE_PCRE2_JIT=1 USE_LINUX_TPROXY=1 USE_LINUX_SPLICE=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_CPU_AFFINITY=1 USE_NS=1 USE_SYSTEMD=1 USE_OT=1 USE_MEMORY_PROFILING=1 USE_MODULES=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT -PCRE2 +PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS +OT -QUIC -PROMEX +MEMORY_PROFILING +MODULES

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with the Prometheus exporter as a service
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Support for malloc_trim() is enabled.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with OpenTracing support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
        [  OT] opentracing

```
```


### Last Outputs and Backtraces

_No response_

### Additional Information

_No response_

**Comment by NickMRamirez** [comment]
Although the table definition requires starting at an index of 1:

```
# setting gpt(0) is an error here, so must set gpt(1)
table response_times type binary len 20 size 1000000 store gpt(1)
```

to access that array value, I have to access it as 0:

```
# does not work
http-response sc-set-gpt(1,0) int(1)

# works
http-response sc-set-gpt(0,0)

# does not work
http-response set-header x-gpt0 %[sc_get_gpt(1,0,mypeers/response_times)]

# works
http-response set-header x-gpt0 %[sc_get_gpt(0,0,mypeers/response_times)]
```

**Comment by NickMRamirez** [comment]
Ultimately, everything works fine when using `gpt0`, but not when using `gpt(0)`. I had trouble with Stick Table Aggregator in HAProxy Enterprise, because although I could hack it to work by offsetting the counter number by one, the aggregated tables would leave the value unset. Reverting to `gpt0` is a work-around.

**Comment by wtarreau** [comment]
Thanks Nick, CCing @EmericBr. It sounds like a mismatch maybe in the parser itself.

**Comment by EmericBr** [comment]
table response_times type binary len 20 size 1000000 store gpt(0)

gpt() : General Purpose Tags Array of elements. This is an array
of positive 32-bit integers which may be used for anything.
Most of the time they will be used to put a special tags on some entries,
for instance to note that a specific behavior was detected and must be
known for future matches. This array is limited to a maximum of 100
elements: gpt0 to gpt99, to ensure that the build of a peer update
message can fit into the buffer.

The argument here is the number of elements of the array to store. An array of 0 elements makes no sense

On table line you are defining a full array, you don't choose counters from an existing array. 

table: array size is out of allowed range (1-100) for store option 'gpt'.

This error message is correct, only arrays of 1 to 100 elements is autorized.

If you want to store only gpc0 you have to define an array of size 1 gpc(1)

If you want to store 100 counters from gpc0 to gpc100 you have to define an array of size 100.

In my point of view there is no bug. We may have to complete the documentation here:
gpt() : General Purpose Tags Array of elements. This is an array
of positive 32-bit integers which may be used for anything.
Most of the time they will be used to put a special tags on some entries,
for instance to note that a specific behavior was detected and must be
known for future matches. This array is limited to a maximum of 100
elements: gpt0 to gpt99, to ensure that the build of a peer update
message can fit into the buffer.

To be more clear that the argument is the number of counters available into the array.

**Comment by EmericBr** [comment]
``
    - gpt(<nb>) : General Purpose Tags Array of <nb> elements. This is an array
      of positive 32-bit integers which may be used for anything.
      Most of the time they will be used to put a special tags on some entries,
      for instance to note that a specific behavior was detected and must be
      known for future matches. This array is limited to a maximum of 100
      elements: gpt0 to gpt99, to ensure that the build of a peer update
      message can fit into the buffer. Users should take in consideration
      that a large amount of counters will increase the data size and the
      traffic load using peers protocol since all data/counters are pushed
      each time any of them is updated.
      This data_type will exclude the usage of the legacy data_type 'gpt0'
      on the same table. Using the 'gpt' array data_type, all 'gpt0' related
      fetches and actions will apply to the first element of this array.

**Comment by EmericBr** [comment]
- gpt(\<nb>) : General Purpose Tags Array of \<nb> elements. This is an array
      of positive 32-bit integers which may be used for anything.
      Most of the time they will be used to put a special tags on some entries,
      for instance to note that a specific behavior was detected and must be
      known for future matches. This array is limited to a maximum of 100
      elements: gpt0 to gpt99, to ensure that the build of a peer update
      message can fit into the buffer. Users should take in consideration
      that a large amount of counters will increase the data size and the
      traffic load using peers protocol since all data/counters are pushed
      each time any of them is updated.
      This data_type will exclude the usage of the legacy data_type 'gpt0'
      on the same table. Using the 'gpt' array data_type, all 'gpt0' related
      fetches and actions will apply to the first element of this array.

**Comment by EmericBr** [comment]
The documentation correctly announce that the parameter is the nb of elements of the array to store (and not an index) your copy pasted has wiped \<nb>

**Comment by EmericBr** [comment]
[0001-DOC-gpc-gpt-add-commments-of-gpc-gpt-array-definitio.patch.txt](https://github.com/haproxy/haproxy/files/8350856/0001-DOC-gpc-gpt-add-commments-of-gpc-gpt-array-definitio.patch.txt)

**Comment by EmericBr** [comment]
I've added some comments and examples to clarify the notion of array size and 0 started indexes

**Comment by EmericBr** [comment]
I've also checked all gpc and gpt keyword occurences and they already clearly expose when the parameter is an index \<idx> or a number of element \<nb>

**Comment by NickMRamirez** [comment]
Ah! Now I understand. The angled brackets around "nb" may have been stripped by the GitHub formatting, not sure. In any case, I think we should play it safe and add this phrase to the definition you wrote: 

`The <nb> parameter indicates the size of the array`.

That should clear up the confusion that this is not indicating an array index.

**Comment by EmericBr** [comment]
Yep, : but the form:  General Purpose Tags Array of \<nb> elements. Is the usual form on the entire haproxy documentation.

Adding  "The <nb> parameter indicates the size of the array (in number of elements) seems quite redundant in my opinion. But if Willy agreed I will add it in the patch.

**Comment by capflam** [comment]
@EmericBr a gentle ping to be sure to fix all doc issue for the 2.6.0 :)

**Comment by wtarreau** [comment]
Closing since this was fixed 2 months ago.
